packages/completer/src/model.ts

```diff
@@ -365,65 +373,106 @@ export class CompleterModel implements Completer.IModel {
    * Check if CompletionItem matches against query.
    * Highlight matching prefix by adding <mark> tags.
    */
-  private _markup(query: string): CompletionHandler.ICompletionItems {
+  private _markup(query: string): {
+    processedItem: CompletionHandler.ICompletionItem;
+    originalItem: CompletionHandler.ICompletionItem;
+  }[] {
     const items = this._completionItems;
     let results: Private.ICompletionMatch[] = [];
-    for (let item of items) {
+    for (const originalItem of items) {
       // See if label matches query string
       // With ICompletionItems, the label may include parameters,
       // so we exclude them from the matcher.
       // e.g. Given label `foo(b, a, r)` and query `bar`,
       // don't count parameters, `b`, `a`, and `r` as matches.
-      const index = item.label.indexOf('(');
-      const text = index > -1 ? item.label.substring(0, index) : item.label;
+      const index = originalItem.label.indexOf('(');
+      const text =
+        index > -1
+          ? originalItem.label.substring(0, index)
+          : originalItem.label;
       const match = StringExt.matchSumOfSquares(escapeHTML(text), query);
       // Filter non-matching items.
       if (match) {
         // Highlight label text if there's a match
         let marked = StringExt.highlight(
-          escapeHTML(item.label),
+          escapeHTML(originalItem.label),
           match.indices,
           Private.mark
         );
         // Use `Object.assign` to evaluate getters.
-        const highlightedItem = Object.assign({}, item);
+        const highlightedItem = Object.assign({}, originalItem);
         highlightedItem.label = marked.join('');
-        highlightedItem.insertText = item.insertText ?? item.label;
+        highlightedItem.insertText =
+          originalItem.insertText ?? originalItem.label;
         results.push({
           item: highlightedItem,
-          score: match.score
+          score: match.score,
+          originalItem
         });
       }
     }
     results.sort(Private.scoreCmp);
 
     // Extract only the item (dropping the extra score attribute to not leak
     // implementation details to JavaScript callers.
-    return results.map(match => match.item);
+    return results.map(match => ({
+      processedItem: match.item,
+      originalItem: match.originalItem
+    }));
   }
 
   /**
-   * Lazy load missing data of item at `activeIndex`.
-   * @param {number} activeIndex - index of item
+   * Lazy load missing data of an item.
+   * @param indexOrValue - the item or its index
+   * @remarks
+   * Resolving item by index will be deprecated in
+   * the next major release.
```

Comment:
Should we be more specific here? Just a suggestion on wording, if you meant other versions that is fine with me too.

```suggestion
   * the JupyterLab 5.0 and removed in JupyterLab 6.0.
```